iwyu.cc

```diff
@@ -1392,12 +1379,23 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
     if (!IsInMacro(use_loc))
       return use_loc;
 
-    // If we're in a macro, and the macro file forward-declares decl, make sure
-    // we keep that forward-declaration.
-    const NamedDecl* fwd_decl = GetForwardDeclareInSameFile(decl, use_loc);
+    // If the macro definition file contains a forward-declaration for decl,
+    // we treat that as a use-attribution hint. See below for details.
+    const FileEntry* macro_def_file = GetFileEntry(use_loc);
+    const NamedDecl* fwd_decl = nullptr;
+    for (const NamedDecl* redecl : GetClassRedecls(decl)) {
+      if (GetLocation(redecl).isFileID() &&
+          GetFileEntry(redecl) == macro_def_file &&
+          IsForwardDecl(redecl)) {
+        fwd_decl = redecl;
+        break;
+      }
+    }
```

Comment:
That one's difficult :-)

The forward-declare rule never comes into play here -- all uses involving macro arguments will be attributed to the expansion location, because, in this case, issue350.cpp is passing in part of the used symbol, so it's assumed to know what it wants to "bind to".

For example, think of how your reasoning would change if it passed in an unknown Logger, e.g.:

    LOG(critical, "test");

Assuming the log framework does not have a `criticalLogger` it's intuitive that the caller is responsible for providing the type.

But as I said, this one is difficult, because your example and conclusion is equally sensible.